eos-esp-generator

```diff
@@ -0,0 +1,968 @@
+
+* Mounting at /boot is allowed even when the /efi directory exists. This allows
+  use of a generic OS commit that can be used on systems where /boot data
+  exists in the ESP or not.
+
+* When the ESP is mounted at /boot, it is made world readable to allow
+  unprivileged ostree admin operations to succeed.
+
+* When the endless.image.device kernel command line argument is set, the
+  ESP from that disk is mounted.
+
+The units created with this generator take precedence over
+systemd-gpt-auto-generator. Output from the generator can be read with
+"journalctl -t eos-esp-generator". When invoked as eos-esp-generator-gather,
+the generator will only gather data and store it in a tarball in /run.
+"""
+
+from argparse import ArgumentParser, RawDescriptionHelpFormatter
+from dataclasses import dataclass
+import json
+import logging
+from logging.handlers import SysLogHandler
+import os
+from pathlib import Path
+import shlex
+import subprocess
+import sys
+import tarfile
+import time
+from tempfile import TemporaryDirectory
+from textwrap import dedent
+
+progname = os.path.basename(__file__)
+logger = logging.getLogger(progname)
+
+ESP_GPT_PARTTYPE = 'c12a7328-f81f-11d2-ba4b-00a0c93ec93b'
+ESP_MBR_PARTTYPE = '0xef'
+XBOOTLDR_GPT_PARTTYPE = 'bc13c2ff-59e6-4262-a352-b275fd6f7172'
+LOADER_EFI_VENDOR = '4a67b082-0a4c-41cf-b6c7-440b29bb8c4f'
+LOADER_DEVICE_PART_UUID_NAME = 'LoaderDevicePartUUID'
+LOADER_DEVICE_PART_UUID_EFIVAR = f'{LOADER_DEVICE_PART_UUID_NAME}-{LOADER_EFI_VENDOR}'
+
+# Program name used to automatically enable gather mode.
+GATHER_MODE_PROGNAME = 'eos-esp-generator-gather'
+
+
+class EspError(Exception):
+    """ESP generator errors"""
+    pass
+
+
+class KmsgHandler(logging.StreamHandler):
+    """Logging Handler using /dev/kmsg"""
+    PRIORITY_MAP = {
+        logging.DEBUG: SysLogHandler.LOG_DEBUG,
+        logging.INFO: SysLogHandler.LOG_INFO,
+        logging.WARNING: SysLogHandler.LOG_WARNING,
+        logging.ERROR: SysLogHandler.LOG_ERR,
+        logging.CRITICAL: SysLogHandler.LOG_CRIT,
+    }
+
+    def __init__(self):
+        self.kmsg = open('/dev/kmsg', 'wb', buffering=0)
+        self.pid = os.getpid()
+        super().__init__(stream=self.kmsg)
+
+    def close(self):
+        self.acquire()
+        try:
+            self.kmsg.close()
+        finally:
+            self.release()
+
+    def emit(self, record):
+        try:
+            message = self.format(record)
+
+            # kmsg recognizes a syslog style <priority>identifier[pid]: prefix.
+            priority = self.PRIORITY_MAP.get(record.levelno, SysLogHandler.LOG_INFO)
+            prefix = f'<{priority:d}>{progname}[{self.pid:d}]: '
+
+            # kmsg allows a maximum of 1024 bytes per message, so split the
+            # message up if needed.
+            size = 1024 - len(prefix)
+            for pos in range(0, len(message), size):
+                buf = f'{prefix}{message[pos:pos + size]}'.encode('utf-8')
+                self.kmsg.write(buf)
```

Comment:
1024 bytes or 1024 characters (which is >= 1024 bytes when encoded as UTF-8)? I assume it's bytes. In that case this will not do the right thing with non-ASCII characters. Consider the message:

```
"☃" * 512
```

Asssuming `prefix` is less than 512 characters, this code will assume that just a single chunk is needed. But:

```
>>> len(("☃" * 512).encode("utf-8"))
1536
```

Unfortunately encoding as UTF-8 before naïvely chunking is also wrong because you can't split the UTF-8-encoded bytes at any arbitrary byte.

I think you need to do something like this:

```python
            message = self.format(record).encode('utf-8')

            # kmsg recognizes a syslog style <priority>identifier[pid]: prefix.
            priority = self.PRIORITY_MAP.get(record.levelno, SysLogHandler.LOG_INFO)
            prefix = f'<{priority:d}>{progname}[{self.pid:d}]: '.encode('utf-8')

            # kmsg allows a maximum of 1024 bytes per message, so split the
            # message up if needed, without splitting multibyte-encoded characters
            start = 0
            length = len(message)
            size = 1024 - len(prefix)
            while start < length:
                end = start + size
                while end < length and (message[end] & 0b1100_0000) == 0b1000_0000:
                    end -= 1

                buf = prefix + message[start:end]
                self.kmsg.write(buf)
```

(The UTF-8 encoding is such that if a byte is of the form `10xxxxxx` it is a continuation byte.)

Reply:
Indeed you are correct. As Aqua once said, I'm an ASCII girl, in an ASCII world. I never understood why UTF-8 was encoded that way, but now I get it! Your suggestion makes perfect sense. Thanks!

Reply:
It's a really clever encoding: the best thing to come out of Plan-9!

Reply:
You algorithm seems to work correctly, although it was missing incrementing `start`. In order to prove it, I added a couple tests. I also added some explicit flushes like `logging.FileHandler` and changed the UTF-8 encoding to use `errors='replace'` so it doesn't bomb on malformed messages.